### PR TITLE
Added Spring-05 hw implementation

### PR DESCRIPTION
**Pull request by oroboro-cpu**

### CODE REVIEW

File: src/main/java/cinema/project/controller/MovieSessionController.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+@RestController
+@RequestMapping("/movie-sessions")
+public class MovieSessionController {
+    private final MovieSessionService movieSessionService;
+    private final MovieSessionMapper movieSessionMapper;
+
+    public MovieSessionController(MovieSessionService movieSessionService,
+                                  MovieSessionMapper movieSessionMapper) {
+        this.movieSessionService = movieSessionService;
+        this.movieSessionMapper = movieSessionMapper;
```
Comment (yurwing) [inline src/main/java/cinema/project/controller/MovieSessionController.java]: add @Autowired

File: src/main/java/cinema/project/controller/MovieSessionController.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        return movieSessionService.findAvailableSessions(movieId, date).stream()
+                .map(movieSessionMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @PutMapping("/{id}")
+    public void update(@RequestBody MovieSessionRequestDto movieSessionRequestDto,
+                       @PathVariable Long id) {
+        MovieSession movieSession = movieSessionMapper.toEntity(movieSessionRequestDto);
+        movieSessionService.update(movieSession);
```
Comment (yurwing) [inline src/main/java/cinema/project/controller/MovieSessionController.java]: you have to set id to movieSession before updating

File: src/main/java/cinema/project/service/mapper/Mapper.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package cinema.project.service.mapper;
+
+public interface Mapper<S, T, P> {
+    T toDto(S t);
+
+    S toEntity(P u);
```
Comment (yurwing) [inline src/main/java/cinema/project/service/mapper/Mapper.java]:
```suggestion
    T toDto(S entity);

    S toEntity(P requestDto);
```

File: src/main/java/cinema/project/service/MovieSessionServiceImpl.java [hunk 4]
```diff
@@ -25,4 +25,19 @@ public MovieSessionServiceImpl(MovieSessionDao movieSessionDao) {
...
+
+    @Override
+    public MovieSession update(MovieSession movieSession) {
+        return movieSessionDao.update(movieSession);
+    }
+
+    @Override
+    public boolean remove(MovieSession movieSession) {
+        return movieSessionDao.remove(movieSession);
+    }
```
Comment (yurwing) [inline src/main/java/cinema/project/service/MovieSessionServiceImpl.java]:
```suggestion
    @Override
    public void remove(MovieSession movieSession) {
        return movieSessionDao.remove(movieSession);
    }
```
in your controller you don't return anything

File: src/main/java/cinema/project/service/MovieSessionServiceImpl.java [hunk 5]
```diff
@@ -25,4 +25,19 @@ public MovieSessionServiceImpl(MovieSessionDao movieSessionDao) {
     public MovieSession add(MovieSession session) {
         return movieSessionDao.add(session);
     }
+
+    @Override
+    public MovieSession update(MovieSession movieSession) {
+        return movieSessionDao.update(movieSession);
+    }
```
Comment (yurwing) [inline src/main/java/cinema/project/service/MovieSessionServiceImpl.java]:
```suggestion
    @Override
    public void update(MovieSession movieSession) {
        return movieSessionDao.update(movieSession);
    }
```
same

**Review by yurwing — CHANGES_REQUESTED** [review]

**Review by yurwing — APPROVED** [review]

File: src/main/java/cinema/project/controller/MovieSessionController.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        return movieSessionService.findAvailableSessions(movieId, date).stream()
+                .map(movieSessionMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @PutMapping("/{id}")
+    public void update(@PathVariable Long id,
+                       @RequestBody MovieSessionRequestDto movieSessionRequestDto) {
+        MovieSession movieSession = movieSessionMapper.toEntity(movieSessionRequestDto);
+        movieSessionService.update(movieSession);
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/project/controller/MovieSessionController.java:57]: should we set the id before calling `update()`?

File: src/main/java/cinema/project/dao/MovieSessionDaoImpl.java [hunk 7]
```diff
@@ -59,4 +60,57 @@ public MovieSession add(MovieSession movieSession) {
...
+    }
+
+    @Override
+    public void remove(Long id) {
+        Session session = null;
+        Transaction transaction = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.delete(get(id));
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/project/dao/MovieSessionDaoImpl.java]: let's not open one session inside of another

File: src/main/java/cinema/project/service/mapper/MovieSessionMapper.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Override
+    public MovieSessionResponseDto toDto(MovieSession movieSession) {
+        MovieSessionResponseDto movieSessionResponseDto = new MovieSessionResponseDto();
+        movieSessionResponseDto.setId(movieSession.getId());
+        movieSessionResponseDto.setMovieId(movieSession.getMovie().getId());
+        movieSessionResponseDto.setCinemaHallId(movieSession.getCinemaHall().getId());
+        movieSessionResponseDto.setShowTime(movieSession.getShowTime()
+                .format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm")));
```
Comment (KseniiaMakarova) [inline src/main/java/cinema/project/service/mapper/MovieSessionMapper.java]: let's make a constant out of this pattern

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**